### PR TITLE
Pass connection information to backend

### LINKED ISSUE #15: Add function to get connection RemoteAddr
Hi,
   it would be useful to have the remote address of the connection to implement a filter based on sender IP.
  I think it could be implemented defining User struct (UserData as User is already ad interface) with RemoteAddr field filled during init of connection.
  
  This change will impact how User is defined as at this moment the Login functions are responsible of User creation. With this change the User object is created by the connection filling all relevant data and passing it to Login functions. Is this doable?

### PR DESCRIPTION
**Pull request by foxcpp**
Will be honest: Didn't checked if it works correctly. But change seems to be pretty trivial so....

Closes #15.

### CODE REVIEW

**Comment by foxcpp** [comment]
I wonder when I will understand that it is a good idea to check changes more carefully before submitting PRs...

File: backend.go [hunk 1]
```diff
@@ -14,11 +14,11 @@ var (
 type Backend interface {
 	// Authenticate a user. Return smtp.ErrAuthUnsupported if you don't want to
 	// support this.
-	Login(username, password string) (User, error)
+	Login(username, password string, info ConnInfo) (User, error)
```
Comment (emersion) [inline backend.go]: Is it possible to make this the first argument instead? And also a pointer?

File: conn.go [hunk 2]
```diff
@@ -160,6 +166,18 @@ func (c *Conn) TLSConnectionState() (state tls.ConnectionState, ok bool) {
 	return tc.ConnectionState(), true
 }
 
+func (c *Conn) Info() (info ConnInfo) {
```
Comment (emersion) [inline conn.go]: Naked returns should be avoided if possible, when they don't convey a value for documentation: https://github.com/golang/go/wiki/CodeReviewComments#named-result-parameters

File: backend.go (hunk 1 above)
Comment (emersion) [reply backend.go]: Okay, not making it a pointer is probably fine, seeing how `tls.ConnectionState` is passed around. Maybe rename to `smtp.ConnectionState` for consistency?

Comment (foxcpp) [reply backend.go]: Sure.

File: conn.go (hunk 2 above)
Comment (foxcpp) [reply conn.go]: Fixed.

File: conn.go [hunk 3]
```diff
@@ -25,6 +25,12 @@ type message struct {
 	To []string
 }
 
+type ConnInfo struct {
```
Comment (emersion) [inline conn.go]: Last one and I stop bothering you: possible to rename to `ConnectionState` for consistency with the `tls` package? :P

Comment (foxcpp) [reply conn.go]: Ok.

File: conn.go [hunk 4]
```diff
@@ -160,6 +166,19 @@ func (c *Conn) TLSConnectionState() (state tls.ConnectionState, ok bool) {
 	return tc.ConnectionState(), true
 }
 
+func (c *Conn) Info() ConnectionState {
```
Comment (emersion) [inline conn.go]: This needs to be changed as well

Comment (foxcpp) [reply conn.go]: How did I... Oops.

**Comment by emersion** [comment]
Thanks!

**Pull request merged**